Add anchors to show/hide sections in install page (#589)
Qiskit/documentation: docs/start/qiskit-versioning.mdx

---
title: Qiskit versioning

> description: Description of how qiskit version numbers work, and information about release schedules, upgrade strategy, pre-releases, and post-releases

Reviewer: Great description!

```suggestion
description: Description of how Qiskit version numbers work, and information about release schedules, upgrade strategy, pre-releases, and post-releases.
```

---

# Qiskit versioning

Qiskit version numbers follow [Semantic Versioning](https://semver.org/).
The version number is comprised of three primary components: the major, minor, and
patch versions. For example, in version number `X.Y.Z`, `X` is the major version,
`Y` is the minor version, and `Z` is the patch version.

Breaking API changes are reserved for major version releases. The **minimum**
period between major version releases is one year. Minor versions introduce
new features and bug fixes without breaking API compatibility, and are
periodically (currently every three months) published for **only** the
current major version. Patch versions provide fixes for bugs identified in
the most recent minor version of each actively supported release series (that is, the 
major version). We support at most two release series at a time, which occurs
only during the period of overlap following a new major version release,
described in more detail below.

## Release schedule

A tentative release schedule is included below:

> ![](/images/start/qiskit-versioning/release_schedule.png)

Reviewer: Thanks for following the image file format <3

are compatible with [PEP440](https://peps.python.org/pep-0440/). Typically
these are release candidates of the form `X.Y.0rc1`. The `rc` releases
will have a finalized API surface and are used to test a prospective release.

Note that when one of the PEP440 pre-release suffixes (such as `a`, `b`, or `pre`) are
published, it does not have the same guarantees as an `rc` release, and is
only a preview release. The API might change between these pre-releases
and the final release with that version number. For example, `1.0.0pre1` might have
a different final API than `1.0.0`.

## Post-releases

If there are issues with a release's packaging, a post-release might be
issued to correct this. These will follow the form `X.Y.Z.1` where the fourth
integer indicates that it is the first post-release of the `X.Y.Z` release.
For example, the qiskit-terra (the legacy package name for Qiskit) 0.25.2
release had some issue with the sdist package publishing, and a post-release
0.25.2.1 was published that corrected this issue. The code was identical, and
0.25.2.1 only fixed the packaging issue for the release.

## Next steps

<Admonition type="tip" title="Recommendations">

>   -  [Select and set up an IBM Quantum channel.](setup-channel)
>   -  [Configure Qiskit locally.](configure-qiskit-local)
>   -  Follow the steps in [Hello world](hello-world) to write and run a quantum program.
>   -  Try one of the [workflow example tutorials.](https://learning.quantum.ibm.com/catalog/tutorials?category=workflow-example)

Author: ```suggestion
  -  [Install and set up Qiskit.](install)
  -  [Configure Qiskit locally.](configure-qiskit-local)
  -  Visit the Qiskit [release notes.](/api/qiskit/release-notes) 
```

---
title: Qiskit versioning
description: Description of how qiskit version numbers work, and information about release schedules, upgrade strategy, pre-releases, and post-releases

---

# Qiskit versioning

Qiskit version numbers follow [Semantic Versioning](https://semver.org/).
The version number is comprised of three primary components: the major, minor, and

> patch versions. For a version number `X.Y.Z` where `X` is the major version,

Reviewer 2: ```suggestion
patch versions. For example, in version number `X.Y.Z`, `X` is the major version,
```

the most recent minor version of each actively supported release series (i.e.
major version). We support at most two release series at a time, which occurs
only during the period of overlap following anew major version release,
described in more detail below.

## Release schedule

A tentative release schedule is included below:

![](/images/start/qiskit-versioning/release_schedule.png)

For an up-to-date release schedule, refer to the Qiskit Github project's
[milestones list](https://github.com/Qiskit/qiskit/milestones), which will always contain the current release plan.

With the release of a new major version, the previous major version is supported
for at least six months; only bug and security fixes will be accepted during this time and only patch releases will be published for this major version. A final
patch version will be published when support is dropped, and that release will
also document the end of support for that major version series. A longer
support window is needed for the previous major version as this gives downstream
consumers of Qiskit a chance to migrate not only their code but also their
users. We do not recommended for a downstream library that

> depends on Qiskit to bump its minimum Qiskit version to a new

Reviewer 2: ```suggestion
depend on Qiskit should not raise their minimum required Qiskit version to a new
```

major version release immediately because its user base will also need a chance
to migrate to the new API changes. Having an extended support window
for the previous major Qiskit version gives downstream projects time to fix
compatibility with the next major version. Downstream projects can provide
support for two release series at a time to give their users a migration path.

For the purposes of semantic versioning, the Qiskit public API is considered
any documented module, class, function, or method that is not marked as private

> (with a `_` prefix). However, there can be explicit exceptions made in the case

Reviewer 2: ```suggestion
(with an underscore `_` prefix). However, there can be explicit exceptions made for
```

> of specific documented APIs. In such cases these APIs will be clearly documented

Reviewer 2: ```suggestion
specific documented APIs. In such cases, these APIs will be clearly documented
```

definition where subclasses are intended to override/implement a private method
as part of defining an implementation of the interface, or advanced-usage
low-level methods that have stable interfaces but are not considered safe to use,
as the burden is on the user to uphold the class/safety invariants themselves
(the canonical example of this is the `QuantumCircuit._append` method).

The supported Python versions, minimum supported Rust version (for building
Qiskit from source), and any Python package dependencies (including the minimum
supported versions of dependencies) used by Qiskit are not part of the backwards
compatibility guarantees and may change during any release. Only minor or major
version releases will raise minimum requirements for using or building Qiskit
(including adding new dependencies), but patch fixes might include support for
new versions of Python or other dependencies. Usually the minimum version of a
dependency is only increased when older dependency versions go out of support or
when it is not possible to maintain compatibility with the latest release of the
dependency and the older version.

## Upgrade strategy

> Whenever a new major version is released, the recommended upgrade path

Reviewer 2: ```suggestion
When a new major version is released, the recommended upgrade path
```

> is to first upgrade to use the most recent minor version on the previous major

Reviewer 2: ```suggestion
is to first upgrade to the most recent minor version on the previous major
```

version. Shortly before a new major version, a final minor version will

> be published. This final minor version release `M.N+1.0` is equivalent to

Reviewer 2: ```suggestion
be published. This final minor version release `X.Y+1.0.0` is equivalent to
```

> `M.N.0` but with warnings and deprecations for any API changes that are

Reviewer 2: ```suggestion
`X.Y.0` but with warnings and deprecations for any API changes that are
```

made on the new major version series.

For example, immediately proceeding the 1.0.0 release, a 0.46.0 release will be
published. The 0.46.0 release will be equivalent to the 0.45.0 release but with
additional deprecation warnings that document the API changes that were made as
part of the 1.0.0 release. This pattern will be used for any future major
version releases.

> As a user of Qiskit, it's recommended that you first upgrade to this final minor

Reviewer 2: ```suggestion
Qiskit users should first upgrade to this final minor
```

> version first, so you can see any deprecation warnings and adjust your Qiskit

Reviewer 2: ```suggestion
version to see any deprecation warnings and adjust their Qiskit
```

> usage ahead of time before trying a potentially breaking release. The previous

Reviewer 2: ```suggestion
usage before trying a potentially breaking release. The previous
```

major version will be supported for at least six months to give sufficient time

> to upgrade. A typical pattern to manage this is to pin the max version to

Reviewer 2: ```suggestion
to upgrade. A typical pattern to manage this is to pin the maximum version to
```

avoid using the next major release series until you're sure of compatibility.

> For example, specifying in a requirements file `qiskit<2` when the current

Reviewer 2: ```suggestion
For example, specifying `qiskit<2` in a requirements file when the current
```

> major Qiskit version is 1 will ensure that you're using a version of Qiskit

Reviewer 2: ```suggestion
major Qiskit version is 1 ensures that you're using a version of Qiskit
```

> that won't have breaking API changes.

Reviewer 2: ```suggestion
that doesn't have breaking API changes.
```

> Pre-emptively capping the version less than the next major version is necessary

Reviewer 2: ```suggestion
Capping the version less than the next major version 
```

> to ensure you get a chance to see deprecation warnings before a

Reviewer 2: ```suggestion
ensures that you see any deprecation warnings before a
```

> major version release. The normal release schedule means the last minor

Reviewer 2: ```suggestion
major version release. 
```

> version release which includes any final deprecation warnings will be released

Reviewer 2: ```suggestion
```

> shortly before the next major version and `pip` will default to using

Reviewer 2: ```suggestion
Without the cap, `pip` installs 
```

> the newest version available unless the version cap is set.

Reviewer 2: ```suggestion
the newest version available by default.
```

## Pre-releases

> For each minor and major version release Qiskit will publish pre-releases that

Reviewer 2: ```suggestion
For each minor and major version release Qiskit publishes pre-releases that
```

are compatible with [PEP440](https://peps.python.org/pep-0440/). Typically
these are release candidates of the form `X.Y.0rc1`. The `rc` releases
will have a finalized API surface and are used to test a prospective release.

Note that when one of the PEP440 pre-release suffixes (such as `a`, `b`, or `pre`) are
published, it does not have the same guarantees as an `rc` release, and is
only a preview release. The API might change between these pre-releases
and the final release with that version number. For example, `1.0.0pre1` might have
a different final API than `1.0.0`.

## Post-releases

> If there are issues with the packaging of a given release, a post-release may be

Reviewer 2: ```suggestion
If there are issues with a release's packaging, a post-release might be
```

issued to correct this. These will follow the form `X.Y.Z.1` where the fourth

> integer is used to indicate it is the first post-release of the `X.Y.Z` release.

Reviewer 2: ```suggestion
integer indicates that it is the first post-release of the `X.Y.Z` release.
```

For example, the qiskit-terra (the legacy package name for Qiskit) 0.25.2
release had some issue with the sdist package publishing, and a post-release
0.25.2.1 was published that corrected this issue. The code was identical, and
0.25.2.1 only fixed the packaging issue for the release.

> ## Next steps

Reviewer 2: I think the only topics that apply as next steps would be Install and set up Qiskit, and Configure Qiskit locally.  And maybe Release notes (https://docs.quantum.ibm.com/api/qiskit/release-notes)
